scvi/model/_condscvi.py

```diff
@@ -92,9 +92,7 @@ def __init__(
 
     @torch.no_grad()
     def get_vamp_prior(
-        self,
-        adata: Optional[AnnData] = None,
-        p: int = 50,
+        self, adata: Optional[AnnData] = None, p: int = 50, resolution: float = 10.0
```

Comment:
10 looks high, and it doesn't guarantee a certain number of clusters, or even the same number of clusters per cell type. How do you think about that? 

kmeans definitely seems like an easier alternative (coding-wise as well). Is the leiden choice justified? 

Also, should we just add the leiden package as an option to scvi-tools instead of using scanpy?

Reply:
agree kmeans would be ideal dependency wise, for now all this is optional dependency.

Reply:
kmeans was my initial choice and I haven't seen performance gain by leiden. Reason to switch were - speed was faster (with 100 initializations for means), selecting n_cluster when there are less cells than clusters is not straightforward, leiden is "default" clustering for single cell.
The resolution is intended to be overclustering (Celltypist uses 5-30 for this purpose https://github.com/Teichlab/celltypist/blob/main/celltypist/classifier.py). I tried multiple values and got closest results to paper with 10 (compared to 5 and 20).
I think new and old code favor cell types with only few cells (sigmas are larger leading to lower loss).

Reply:
can't you use some kind of other spectral clustering or minibatch kmeans in sklearn?

Reply:
It doesn't really depend on the current implementation. I will update it most likely tomorrow (need to run the three datasets again).

Reply:
List was slower. However, now it uses a tensor with shape kmeans number of clusters, uses cluster probability as weights and sets values to a weight of 0 if less clusters are given for the corresponding cell type.